### ISSUE TITLE
The state of dev branch?

### ISSUE THREAD

**Issue by robinwassen** [mathquill/mathquill#440]
When is the dev branch considered stable enough for being merged into master? Or what is considered not done in the dev branch?

**Comment by laughinghan** [comment]
Good question. I need a centralized place this is written down, I'm not 100% sure. The 2 big things in my mind are:
- writing a changelog, with detailed explanations of how to port from the old jQuery plugin-based API to the new one
- decide whether to make even more backcompat-breaking changes, especially to deal with the utter dearth of error reporting when parsing LaTeX

**Comment by robinwassen** [comment]
I would recommend that you go easy on yourself and release a 1.0.0 version of the library - if you want to do some major breaking changes in the future you can save those for a 2.0.0 release instead.

When it comes to the change log I would say that if you bring up the most major changes (e.g. no jQuery dependency etc) and refer to some docs on how to use the new version it's probably enough. And for details you can just refer to the pull request for dev -> master.

This way you can make the library move forward faster.

**Comment by EvilOatmeal** [comment]
I'm working together with @robinwassen at DigiExam and we want to make some major contributions to MathQuill (for example predefined values in commands (issue #69) and an undo/redo manager (issue #5)). We also want many of the smaller fixes already done in forks but that have not been merged back into the main repository.

However if we begin doing this work in our own fork now we feel that the project would become even more fragmented and we would have to work with unnecessary complexity.

This comment is an outreach to the maintainers and major contributors of MathQuill to set a clear target for MathQuill, release a new version, and then get fixes and features from the forks back into the main repository.

I suggest these goals for now:

1. Create a GitHub issues milestone and collect everything that needs to be done before the "dev" branch can be released
2. Write a changelog with a migration guide
3. Release the "dev" branch
4. Clean up this repository in terms of branches, issues and documentation (readme and wiki)
5. Involve the major contributors to MathQuill and try to merge their fixes and features back into the main repository
  * Desmos: @eluberoff, @jwmerrill 
  * Learnosity: @eoghanmcilwaine, @karoltarasiuk 
  * Khan: @alopatin
  * @brentan
  * @taylor-work

MathQuill seems to be the only major JS library when it comes to writing math so it's important that we clean things up and make contributing to it an easy process.

How do you other contributors feel about this? Could we get a collaboration started?

**Comment by laughinghan** [comment]
I agree completely with everything you've said. :)

Tonight, I'll create a milestone for releasing "dev" and organize the remaining issues I want resolved before that happens.

**Comment by brentan** [comment]
I would be open to contributing and agree a less fragmented version would
be nice...I had originally started updating and modifying mathquill for my
purposes using extra options and flags, but quickly abandoned that effort
and just went my own way with my branch because:

   - As you note, the mathquill world was already fragmented
   - I felt my goals/needs were different than mathquill's original design
   intent, and so it seemed the changes I was making were of a major variety
   (directional changes, not fixes)
      - For example, I am using mathquill as an I/O display engine for a
      web-based calculator (http://www.swiftcalcs.com, invite code is
      'alpha')....latex support isn't terribly useful to me, but 'computable'
      text input/output is (for example, instead of \frac{x^{2+y}}{2}} as the
      parse strings from mathquill, I want/need 'x^(2+y)/2'. That means I have
      broken latex support throughout as I change parser rules and output to
      produce the format I need. No reason these couldn't co-exist, I just
      haven't cared to keep pure latex support functional.
      - Another example: I have stripped out the textarea and all the
      keyboard/mouse listener stuff that goes along with that, and I
handle that
      external to mathquill in my own app. I simply added a few API
functions to
      send keypresses and mouse events to mathquill.
   - This is a side project for me, so I haven't always followed good
   coding practices...

For me, I think an important first step would be for the community to reach
a consensus on the goals of mathquill and what features/options we all want
to see on the roadmap. If that hits all of our needs, there is a clear
pathway to merge various fixes and things together. If not, I am happy to
contribute fixes and things to a master branch when we find/fix them, but
we will probably always need a separate branch focused on our own needs.

Best,
Brentan

**Comment by karoltarasiuk** [comment]
Completely agree with Kenneth on behalf of Learnosity. It's a great project
and we should definitely keep it rolling. As Brentan has mentioned some of
the goals of forked repos can be different. Different are also contexts
we're using it in. But it doesn't mean we can't bring some changes back to
official repo if it meets the roadmap of creators, which I'm more than
happy to do.

I'd suggest one more thing - documenting the code better, and also writing
some knowledgebase article about the general structure and components of
the project. I still struggle with some of it, and it's not that easy to
understand everything from the code. Having an article would lower down the
barrier for new developers who want to help.

2015-08-07 6:59 GMT+10:00 Brentan Alexander <notifications@github.com>:
